### DISCUSSION TITLE
Building OpenSSL with "Configure no-shared" still tries to build  shared libraries

### DISCUSSION THREAD

**Discussion by winspool** [openssl/openssl#23901] [Q&A] [answered]
Another issue detected, while try to build OpenSSL with `tcc`:

OpenSSL version:
```
commit dc9bc6c8e1bd329ead703417a2235ab3e97557ec (HEAD -> master, origin/openssl-3.3, origin/master, origin/HEAD)
```

Commandline used for Configuration:
```
../Configure no-shared no-asm

i also tried with:
../Configure no-shared no-asm linux-generic64

```

Build output:
```
tcc -fPIC -pthread -Wall -O3 -L. -Wl,-z,defs -Wl,-znodelete -shared -Wl,-Bsymbolic   \
	-o providers/legacy.so -Wl,--version-script=providers/legacy.ld \
	providers/legacy-dso-legacyprov.o \
	providers/liblegacy.a providers/libcommon.a -lcrypto -ldl -pthread 
tcc: error: unsupported linker option '-znodelete'
make[1]: *** [Makefile:6806: providers/legacy.so] Error 1
make[1]: Leaving directory '/data/_s/ssh/openssl/_build.tcc'
```

I know, that `-Wl,-znodelete` is an important security feature,
which is currently not supported by tcc
(and tcc also does not support `-Wl,--version-script=providers/legacy.ld` yet),
but the issue here is that `no-shared` was given
and OpenSSL still tries to build the shared library `providers/legacy.so`.

When using gcc (13.2.0), the following shared libraries are build with `Configure no-shared`:
```
1169:gcc -fPIC -pthread -Wall -O3 -L. -Wl,-z,defs -Wl,-znodelete -shared -Wl,-Bsymbolic   \
1170-	-o providers/legacy.so -Wl,--version-script=providers/legacy.ld \
1171-	providers/legacy-dso-legacyprov.o \
1172-	providers/liblegacy.a providers/libcommon.a -lcrypto -ldl -pthread 
--
1175:gcc -fPIC -pthread -Wall -O3 -Wl,-z,defs -Wl,-znodelete -shared -Wl,-Bsymbolic   \
1176-	-o test/p_minimal.so -Wl,--version-script=test/p_minimal.ld \
1177-	test/p_minimal-dso-p_minimal.o \
1178-	-ldl -pthread 
--
1181:gcc -fPIC -pthread -Wall -O3 -Wl,-z,defs -Wl,-znodelete -shared -Wl,-Bsymbolic   \
1182-	-o test/p_test.so -Wl,--version-script=test/p_test.ld \
1183-	test/p_test-dso-p_test.o \
1184-	-ldl -pthread 
```

I tested with "no-shared", because OpenWatcom can`t build shared libraries on linux (yet).

**Comment by bernd-edlinger** [comment]
can you try `no-pic no-dso` instead?

**Comment by levitte** [comment]
We distinguish between "shared libraries" (shared objects that the programs are linked with at build time) and "modules" (shared objects that are linked to at run time).  `no-shared` affects only the former, and to disable the build of dynamically loaded modules, you must also specify `no-module`.

Note that @bernd-edlinger's alternatives aren't wrong either.  `no-pic` implies that no shared object will be built (it is literally an alias for `no-shared no-module`), and `no-dso` is essentially an (older) alias for `no-module`.

**Reply by winspool** [reply]
Thanks for the hints.

Building OpenSSL is new for me 
and a pearl script with make as a build system
looks very strange to me as newcomer.
(Additionally, I do not know pearl).

**Comment by lytex** [comment]
I'd like to build openssl statically too, but I'm still getting libcrypt and libssl even after enabling all the options listed. Also I pass `-static` and `--static` to gcc. Is there something I'm missing?

```dockerfile
FROM debian:bookworm-20240311

RUN mkdir -p /app
WORKDIR /app
RUN apt-get update && apt-get install -y git gcc make && rm -rf /var/lib/apt/lists/* /var/cache/apt/*
RUN git clone https://github.com/openssl/openssl
WORKDIR /app/openssl
RUN ./Configure no-docs no-shared no-pinshared no-module no-pic no-dso no-dynamic-engine -static --static
RUN make
RUN make install

ENTRYPOINT ["/usr/bin/ldd", "/usr/bin/openssl"]
```

```
docker build . -t openssl

docker run openssl
        linux-vdso.so.1 (0x00007ffe621ec000)
        libssl.so.3 => /lib/x86_64-linux-gnu/libssl.so.3 (0x000078a5a477b000)
        libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x000078a5a42f9000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000078a5a4118000)
        /lib64/ld-linux-x86-64.so.2 (0x000078a5a491b000)
```

The commit is 1967539e212c17139dc810096da987c8100b1ba2

**Reply by kulkarniamit** [reply, marked as answer]
According to [openssl INSTALL.md](https://github.com/openssl/openssl/blob/master/INSTALL.md#unix--linux--macos--nonstop-1) if you don't use `--prefix` option with `Configure`, By default, OpenSSL will be installed to
```
/usr/local
```
More precisely, the files will be installed into the subdirectories
```
/usr/local/bin
/usr/local/lib
/usr/local/include
...
```
So, you may want to edit the last line to:
```
ENTRYPOINT ["/usr/bin/ldd", "/usr/local/bin/openssl"]
```

**Reply by lytex** [reply]
Yup, working now.

I have an app at /usr/bin/openssl still, it was not present on the base image. ~~Does the build create a /usr/bin/openssl with dynamic libraries and a static /usr/local/bin/openssl ?~~ Probably came as a dependency from git, to work with https repos
